ui/ts/pages/sql.ts

```diff
@@ -0,0 +1,113 @@
+      import MithrilBasicProperty = _mithril.MithrilBasicProperty;
+      import TableData = Components.Table.TableData;
+      import Column = Models.Proto.Column;
+      import Row = Models.Proto.Row;
+      import Response = Models.Proto.Response;
+      import TableColumn = Components.Table.TableColumn;
+      import MithrilController = _mithril.MithrilController;
+      import Datum = Models.Proto.Datum;
+      import Property = Utils.Property;
+
+      function runQuery(q: string, data: MithrilBasicProperty<Response>): void {
+        Models.SQLQuery.runQuery(q).then(data);
+      }
+
+      function populateTableDataFromResult(result: Response, table: TableData<Row>): void {
+        let cols: Column[] = _.get(result, "results[0].Union.Rows.columns", []);
+        let resultRows: Row[] = _.get(result, "results[0].Union.Rows.rows", []);
+        let tableColumns = <Property<TableColumn<Row>[]>>table.columns;
+        tableColumns(_.map(cols, function (col: Column, i: number): TableColumn<Row> {
+            return {
+              title: col.name,
+              view: function (row: Row): string {
+                let payload: Datum = row.values[i].Payload;
+                let type: string = _.keys(payload)[0];
+                let viewVal: string = "";
+                if (type === "BytesVal") {
+                  viewVal = payload.BytesVal.toString();
+                } else if (type === "TimeVal") {
+                  viewVal = Utils.Convert.TimestampToMoment(payload.TimeVal).toString();
+                } else if (type === "DateVal") {
+                  viewVal = moment.utc(0).add(payload.DateVal, "days").format("YYYY-MM-DD");
+                } else {
+                  viewVal = payload[type].toString();
+                }
+                return viewVal;
+              },
+            };
+          }));
+
+        (<Property<Row[]>>table.rows)(resultRows);
+      }
+
+      class SQLController implements MithrilController {
+        displayJSON: boolean = false;
+        data: MithrilBasicProperty<Response> = m.prop({});
+        columns: Property<TableColumn<Row>[]> = Utils.Prop([]);
+        rows: Property<Row[]> = Utils.Prop([]);
+
+        tableData: TableData<Row> = {
+          columns: this.columns,
+          rows: this.rows,
+        };
+
+        query: MithrilBasicProperty<string> = m.prop("");
+
+        toggleDisplayJSON(): void {
+          this.displayJSON = !this.displayJSON;
+        };
+        toggleText(): string {
+          return this.displayJSON ? "Hide JSON Response" : "Show JSON Response";
+        };
+      }
+
+      export function controller(): SQLController {
+        let ctrl: SQLController = new SQLController();
+        ctrl.query(m.route.param("q"));
+        runQuery(ctrl.query(), ctrl.data);
+        return ctrl;
+      }
+
+      export function view(ctrl: SQLController): MithrilVirtualElement {
+
+        populateTableDataFromResult(ctrl.data(), ctrl.tableData);
```

Comment:
This kind of thing should only be run when the underlying query data changes, not every time the view is updated.  I think in this case you could move it to the controller method (since the query is only run one time), but another alternative would be to use one of our computed properties.  You can dump the output of runQuery into one of our properties instead of a mithril property (they are compatible) and then make "populatedTableData" a computed property instead of a function.  That way, populatedTableData will only be recomputed if the query data changes.